### ISSUE TITLE
Providing a parameter value from middleware

### ISSUE THREAD

**Issue by NibblyPig** [Azure/azure-functions-dotnet-worker#1300]
Using Middleware in an azure function, is it possible to modify the context to bind an object to a parameter in the function?

```
public class FunctionInvocationLoggingMiddleware : IFunctionsWorkerMiddleware
    {
        public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
        {
            context.SupplyParameter("myObject", new someObject());
            await next(context);
        }
    }
```

```
[Function("Function1")]
        public HttpResponseData Run([HttpTrigger(AuthorizationLevel.Function, "get", "post")] HttpRequestData req,
            someObject myObject) // This value is provided by the middleware if requested
        {
            ....
        }
```

This would be similar to MVC value providers, I guess.

The use case is that I have a lot of functions that take a blob of xml, and it would be nice to be able to parse, log and debug that xml in the middleware before simply providing the C# type to the function all nice and ready to go.

**Comment by SeanFeldman** [comment]
Yes, it is possible. You could add your object(s) to `FunctionContext ` via `context.Items.Add()` and retrieve it in your function `FunctionContext` by using `FunctionContext.Items`. You will need to inject the context into your function by passing it as a parameter.

```c#
[Function("Function1")]
public HttpResponseData Run([HttpTrigger(AuthorizationLevel.Function, "get", "post")] HttpRequestData req, FunctionContext context)
{
	SomeObject myObject = (SomeObject)context.Items["key"];
}
```

In your middleware:

```c#
public class FunctionInvocationMiddleware : IFunctionsWorkerMiddleware
{
  public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
  {
      context.Items.Add("key", new someObject());
      await next(context);
  }
}
```

**Comment by NibblyPig** [comment]
Given that it can populate things without any of your own code, e.g. HttpRequestData and now FunctionContext, is there no way to insert your object and have it autopopulate just like those do via some kind of provider?

If not, I think I can make this work. I appreciate your response!

**Comment by SeanFeldman** [comment]
I only know of route parameters binding and the context bag.
Perhaps @jviau could point out the documentation on anything existing or issues tracking requests for what you're asking for.

**Comment by fabiocav** [comment]
@NibblyPig / @SeanFeldman this is possible today, and the good news is that you've implemented 50% of the required code here already. 

To tie this together, you'd be required to implement an input converter. We don't have great documentation on that yet, so I'm tagging @kshyju to share some guidance.

**Comment by kshyju** [comment]
@NibblyPig / @SeanFeldman While we are working on getting the official docs updated, please take a look at [Extending input binding pipeline in dotnet isolated Azure functions](https://techiesweb.net/2023/02/11/azure-functions-input-converters.html)

Hope this helps. Don't hesitate to reach out if you run into issues. Thanks!

**Comment by SeanFeldman** [comment]
@kshyju it does. And the information in the post would be perfect documentation material. Thank you.

**Comment by fabiocav** [comment]
Moving this to sprint 140 as a documentation item (@cloudmelon for awareness) and to track a sample implementation in the repo.

**Comment by eswann** [comment]
Hey all, I was trying to do the same thing here. I have existing code that already uses the `ConfigureFunctionsWorkerDefaults` with the HostBuilder action instead of the WorkerOptions action. Also, this is a function triggered by a Storage Queue binding, not HTTP. When I attempt to add in my InputConverter, everything compiles but my function is never called. I'm running locally with the storage emulator BTW. The HostBuilder code is pasted below. Should this work? If I instead just add the InputConverter registration into the "builder" block, it won't compile, giving me an "ambiguous invocation" compiler error.
```
        host = new HostBuilder()
            .ConfigureFunctionsWorkerDefaults
            (
                workerOptions => { workerOptions.InputConverters.Register<CustomMessageConverter>(); }
            )
            .ConfigureFunctionsWorkerDefaults
            (
                builder =>
                {
                    builder.ConfigureSystemTextJson();
                    builder.UseFunctionContextAccessor();
                    builder.UseCorrelationIdProviderMiddleware();
                    builder.UseTenantProviderMiddleware();
                }
            ) 
```

**Comment by fabiocav** [comment]
Moving this to triage so we can create the doc tracking items.

**Comment by fabiocav** [comment]
This doc issue was assigned to a stale milestone. Flagging for triage.

**Comment by thedomeffm** [comment]
> Hey all, I was trying to do the same thing here. I have existing code that already uses the `ConfigureFunctionsWorkerDefaults` with the HostBuilder action instead of the WorkerOptions action. Also, this is a function triggered by a Storage Queue binding, not HTTP. When I attempt to add in my InputConverter, everything compiles but my function is never called. I'm running locally with the storage emulator BTW. The HostBuilder code is pasted below. Should this work? If I instead just add the InputConverter registration into the "builder" block, it won't compile, giving me an "ambiguous invocation" compiler error.
> 
> ```
>         host = new HostBuilder()
>             .ConfigureFunctionsWorkerDefaults
>             (
>                 workerOptions => { workerOptions.InputConverters.Register<CustomMessageConverter>(); }
>             )
>             .ConfigureFunctionsWorkerDefaults
>             (
>                 builder =>
>                 {
>                     builder.ConfigureSystemTextJson();
>                     builder.UseFunctionContextAccessor();
>                     builder.UseCorrelationIdProviderMiddleware();
>                     builder.UseTenantProviderMiddleware();
>                 }
>             ) 
> ```

@eswann 
I am running into the same problem like you. Did you find a solution for registering the InputConverter?
I didn't find a documentation with an example.

**Comment by kshyju** [comment]
@thedomeffm Can you please open a new issue with relevant code which can be used to repro your issue?

**Comment by mattchenderson** [comment]
The two calls approach may actually work now when referencing the V2 worker and SDK packages. But another option would be to configure the `WorkerOptions` using the service collection exposed by the builder itself.

```csharp
var host = new HostBuilder()
    .ConfigureFunctionsWorkerDefaults((IFunctionsWorkerApplicationBuilder builder) =>
    {
        builder.Services.Configure<WorkerOptions>(workerOptions =>
        {
            workerOptions.InputConverters.Register<CustomMessageConverter>();
        });
        builder.ConfigureSystemTextJson();
        builder.UseFunctionContextAccessor();
        builder.UseCorrelationIdProviderMiddleware();
        builder.UseTenantProviderMiddleware();
    })
    .Build();

host.Run();
```

This is perhaps a bit cleaner when using the new `IHostApplicationBuilder` model that those newer package versions add:

```csharp
var builder = FunctionsApplication.CreateBuilder(args);

builder.Services.Configure<WorkerOptions>(workerOptions =>
{
    workerOptions.InputConverters.Register<CustomMessageConverter>();
});

builder.ConfigureSystemTextJson();
builder.UseFunctionContextAccessor();
builder.UseCorrelationIdProviderMiddleware();
builder.UseTenantProviderMiddleware();

builder.Build().Run();

```
